standard/classes.md

```diff
@@ -5584,3 +5466,146 @@ When the body of the async function terminates, the return task is moved out of
+
+- To advance the enumerator: `MoveNext` and `MoveNextAsync`.
+- To retrieve the current value: `Current`.
+- To dispose of resources: `Dispose` and `DisposeAsync`.
+
+Enumerator objects do not support the `IEnumerator.Reset` method. Invoking this method causes a `System.NotSupportedException` to be thrown.
+
+Synchronous and asynchronous iterator blocks differ in that asynchronous iterator members return task types and may be awaited.
+
+#### 15.14.5.2 Advance the enumerator
+
+The `MoveNext` and `MoveNextAsync` methods of an enumerator object encapsulates the code of an iterator block. Invoking the `MoveNext` or `MoveNextAsync` method executes code in the iterator block and sets the `Current` property of the enumerator object as appropriate.
+
+`MoveNext` returns a `bool` value whose meaning is described below. `MoveNextAsync` returns a `ValueTask<bool>` (§15.15.3). The result value of the task returned from `MoveNextAsync` has the same meaning as the result value from `MoveNext`. In the following description, the actions described for `MoveNext` apply to `MoveNextAsync` with the following difference: Where stated that `MoveNext` returns `true` or `false`, `MoveNextAsync` sets its task to the *completed* state, and sets the task's result value to the corresponding `true` or `false` value.
+
+The precise action performed by `MoveNext` or `MoveNextAsync` depends on the state of the enumerator object when invoked:
+
+- If the state of the enumerator object is **before**, invoking `MoveNext`:
+  - Changes the state to **running**.
+  - Initializes the parameters (including `this`) of the iterator block to the argument values and instance value saved when the enumerator object was initialized.
+  - Executes the iterator block from the beginning until execution is interrupted (as described below).
+- If the state of the enumerator object is **running**, the result of invoking `MoveNext` is unspecified.
+- If the state of the enumerator object is **suspended**, invoking MoveNext:
+  - Changes the state to **running**.
+  - Restores the values of all local variables and parameters (including `this`) to the values saved when execution of the iterator block was last suspended.  
+    > *Note*: The contents of any objects referenced by these variables may have changed since the previous call to `MoveNext`. *end note*
+  - Resumes execution of the iterator block immediately following the yield return statement that caused the suspension of execution and continues until execution is interrupted (as described below).
+- If the state of the enumerator object is **after**, invoking `MoveNext` returns false.
+
+When `MoveNext` executes the iterator block, execution can be interrupted in four ways: By a `yield return` statement, by a `yield break` statement, by encountering the end of the iterator block, and by an exception being thrown and propagated out of the iterator block.
```

Comment:
Do we need to add a fifth way for asynchronous iterators, namely "awaiting something awaitable that hasn't completed yet"?

Reply:
I added a note in #672, because that can be derived from the fact that `MoveNextAsync` may include `await` expressions.